src/videoframeprocessor/DefaultVideoTransformDevice.ts

```diff
@@ -0,0 +1,224 @@
+
+/**
+ * [[DefaultVideoTransformDevice]] is an augmented [[VideoInputDevice]].
+ * It transform the input {@link Device} with an array of {@link VideoFrameProcessor} to produce a `MediaStream`.
+ */
+export default class DefaultVideoTransformDevice
+  implements VideoTransformDevice, VideoFrameProcessorPipelineObserver {
+  private inputMediaStream: MediaStream;
+  private observers: Set<DefaultVideoTransformDeviceObserver> = new Set<
+    DefaultVideoTransformDeviceObserver
+  >();
+
+  constructor(
+    private logger: Logger,
+    private device: Device,
+    private processors: VideoFrameProcessor[],
+    private browserBehavior: DefaultBrowserBehavior = new DefaultBrowserBehavior(),
+    private pipe: DefaultVideoFrameProcessorPipeline | null = null
+  ) {}
+
+  /**
+   * getter for `outputMediaStream`.
+   * `outputMediaStream` is returned by internal {@link VideoFrameProcessorPipeline}.
+   */
+  get outputMediaStream(): MediaStream | null {
+    if (this.pipe) {
+      return this.pipe.outputMediaStream;
+    }
+    return null;
+  }
+
+  /**
+   * `chooseNewInnerDevice` preserves the inner pipeline and processing state and switches
+   * the inner device. Since the pipeline and processors are shared with the new transform device
+   * only one transform device can be used.
+   */
+  chooseNewInnerDevice(newDevice: Device): DefaultVideoTransformDevice {
+    const newTransformDevice = new DefaultVideoTransformDevice(
+      this.logger,
+      newDevice,
+      this.processors,
+      this.browserBehavior,
+      this.pipe
+    );
+    return newTransformDevice;
+  }
+
+  /**
+   * Return the inner device as provided during construction.
+   */
+  getInnerDevice(): Device {
+    return this.device;
+  }
+
+  async intrinsicDevice(): Promise<Device> {
+    const trackConstraints: MediaTrackConstraints = {};
+
+    // Empty string and null.
+    if (!this.device) {
+      return trackConstraints;
+    }
+
+    // Device ID.
+    if (typeof this.device === 'string') {
+      if (this.browserBehavior.requiresNoExactMediaStreamConstraints()) {
+        trackConstraints.deviceId = this.device;
+      } else {
+        trackConstraints.deviceId = { exact: this.device };
+      }
+      return trackConstraints;
+    }
+
+    if ((this.device as MediaStream).id) {
+      // Nothing we can do.
+      return this.device;
+    }
+
+    // It's constraints.
+    return {
+      ...this.device,
+      ...trackConstraints,
+    };
+  }
+
+  /**
+   * Create {@link VideoFrameProcessorPipeline} if there is not a existing one and start video processors.
+   * Returns output `MediaStream` produced by {@link VideoFrameProcessorPipeline}.
+   */
+  async applyProcessors(mediaStream?: MediaStream): Promise<MediaStream> {
+    await this.createVideoFrameProcessorPipeline();
+    await this.pipe.setInputMediaStream(mediaStream);
+    this.inputMediaStream = mediaStream;
+    return this.pipe.outputMediaStream;
+  }
+
```

Comment:
If the pipeline is empty, will it pass through the stream unmolested (e.g., defaulting to a `NoOp` step)?

If the pipeline is absent (I suggest you prevent that case by not having it as an argument!), can you make `DVTD` just pass through the `mediaStream` here?

Reply:
Yes, if the pipeline is empty, it will pass through the stream like a no op step.

I move the pipeline out of argument. And `DVTD` should always have a pipeline if it is to be used by device controller